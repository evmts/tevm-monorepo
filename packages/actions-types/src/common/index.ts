export type { Abi } from './Abi.js'
export type { Block } from './Block.js'
export type { BlockParam } from './BlockParam.js'
export type { BlockTag } from './BlockTag.js'
export type { Address } from './Address.js'
export type { Hex } from './Hex.js'
export type { EmptyParams } from './EmptyParams.js'
export type { Log } from './Log.js'
export type { BlockResult } from './BlockResult.js'
export type { FilterLog } from './FilterLog.js'
export type { FilterParams } from './FilterParams.js'
export type { TransactionParams } from './TransactionParams.js'
export type { TransactionResult } from './TransactionResult.js'
export type { TransactionReceiptResult } from './TransactionReceiptResult.js'
export type { TraceCall } from './TraceCall.js'
export type { TraceResult } from './TraceResult.js'
export type { TraceType } from './TraceType.js'
export type { NetworkConfig } from './NetworkConfig.js'
