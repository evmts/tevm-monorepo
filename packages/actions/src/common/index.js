export { isValidCallEvents, validateCallEvents } from './zCallEvents.js'
