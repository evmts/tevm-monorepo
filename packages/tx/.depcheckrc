ignores: [
  "@ethereumjs/block",
  "@ethereumjs/common"
]
skip-missing: true
