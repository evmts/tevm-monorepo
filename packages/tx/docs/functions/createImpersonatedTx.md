[**@tevm/tx**](../README.md) • **Docs**

***

[@tevm/tx](../globals.md) / createImpersonatedTx

# Function: createImpersonatedTx()

> **createImpersonatedTx**(`txData`, `opts`?): [`ImpersonatedTx`](../interfaces/ImpersonatedTx.md)

## Parameters

• **txData**: `FeeMarketEIP1559TxData` & `object`

• **opts?**: [`TxOptions`](../interfaces/TxOptions.md)

## Returns

[`ImpersonatedTx`](../interfaces/ImpersonatedTx.md)

## Source

[packages/tx/src/createImpersonatedTx.js:9](https://github.com/evmts/tevm-monorepo/blob/main/packages/tx/src/createImpersonatedTx.js#L9)
