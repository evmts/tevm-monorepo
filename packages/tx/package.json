{
  "name": "@tevm/tx",
  "version": "2.0.0-next.105",
  "private": false,
  "description": "A custom implementation of ethereumjs tx",
  "keywords": [
    "solidity",
    "ethereumjs",
    "typescript",
    "web3",
    "blockchain"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/evmts/tevm-monorepo.git",
    "directory": "packages/tx"
  },
  "license": "MIT",
  "contributors": [
    "Will Cory <willcory10@gmail.com>"
  ],
  "type": "module",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./types/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "types/index.d.ts",
  "files": [
    "dist",
    "types",
    "src",
    "!src/**/*.spec.ts"
  ],
  "scripts": {
    "all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs",
    "build": "nx run-many --targets=build:dist,build:types --projects=@tevm/tx",
    "build:dist": "tsup",
    "build:types": "tsup --dts-only && tsc --emitDeclarationOnly --declaration",
    "clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
    "format": "biome format . --write",
    "format:check": "biome format .",
    "generate:docs": "typedoc",
    "lint": "biome check . --write --unsafe",
    "lint:check": "biome check . --verbose",
    "lint:deps": "bunx depcheck",
    "lint:package": "bunx publint --strict && attw --pack",
    "package:up": "pnpm up --latest",
    "test": "vitest --watch",
    "test:coverage": "vitest --coverage",
    "test:run": "vitest",
    "typecheck": "tsc --noEmit"
  },
  "dependencies": {
    "@ethereumjs/tx": "^5.3.0",
    "@tevm/errors": "workspace:^",
    "@tevm/utils": "workspace:^"
  },
  "devDependencies": {
    "@ethereumjs/common": "^4.3.0",
    "@tevm/tsconfig": "workspace:^",
    "@tevm/tsupconfig": "workspace:^"
  },
  "publishConfig": {
    "access": "public"
  },
  "sideEffect": false
}
