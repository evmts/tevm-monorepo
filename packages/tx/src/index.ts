export {
	type EIP1559CompatibleTx,
	type TxData,
	type TypedTransaction,
	LegacyTransaction,
	AccessListEIP2930Transaction,
	FeeMarketEIP1559Transaction,
	AccessListEIP2930Transaction,
	BlobEIP4844Transaction,
	TransactionFactory,
	Capability,
	isAccessListEIP2930Tx,
	isBlobEIP4844Tx,
	isFeeMarketEIP1559Tx,
	isLegacyTx,
} from '@ethereumjs/tx'
