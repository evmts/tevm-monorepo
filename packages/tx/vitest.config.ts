import { defineConfig } from 'vitest/config'

// https://vitest.dev/config/ - for docs
export default defineConfig({
	test: {
		include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
		environment: 'node',
		coverage: {
			include: ['src/**/*.js'],
			provider: 'v8',
			reporter: ['text', 'json-summary', 'json'],
			thresholds: {
				autoUpdate: true,
				// Lower thresholds temporarily to accommodate changes for ethereumjs alpha
				lines: 93.42,
				functions: 100,
				branches: 95.83,
				statements: 93.42,
			},
		},
	},
})
