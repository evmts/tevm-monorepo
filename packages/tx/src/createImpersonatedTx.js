import { FeeMarket1559Transaction } from '@ethereumjs/tx'
import { InternalError, InvalidGasLimitError } from '@tevm/errors'
import { EthjsAddress, keccak256 } from '@tevm/utils'

/**
 * @typedef {InternalError | InvalidGasLimitError} CreateImpersonatedTxError
 */

/**
 * Creates an impersonated tx that wraps {@link FeeMarket1559Transaction}.
 * Wraps following methods
 * - 'isImpersonated'
 * - 'hash'
 * - 'isSigned'
 * - 'getSenderAddress'
 * @throws {CreateImpersonatedTxError} Error if the constructor for {@link FeeMarket1559Transaction} throws
 * @param {import("@ethereumjs/tx").FeeMarketEIP1559TxData & {impersonatedAddress: EthjsAddress}} txData
 * @param {import("@ethereumjs/tx").TxOptions} [opts]
 * @returns {import('./ImpersonatedTx.js').ImpersonatedTx}
 */
export const createImpersonatedTx = (txData, opts) => {
	/**
	 * @type {FeeMarket1559Transaction}
	 */
	let tx
	try {
		tx = new FeeMarket1559Transaction(txData, opts)
	} catch (e) {
		if (!(e instanceof Error)) {
			throw new InternalError('Unknown Error', { cause: /** @type any*/ (e) })
		}
		if (e.message.includes('EIP-1559 not enabled on Common')) {
			throw new InternalError(
				'EIP-1559 is not enabled on Common. Tevm currently only supports 1559 and it should be enabled by default',
				{ cause: e },
			)
		}
		if (
			e.message.includes('gasLimit cannot exceed MAX_UINT64 (2^64-1)') ||
			e.message.includes('gasLimit * maxFeePerGas cannot exceed MAX_INTEGER (2^256-1)')
		) {
			throw new InvalidGasLimitError(e.message, { cause: e })
		}
		if (
			e.message.includes(
				'maxFeePerGas cannot be less than maxPriorityFeePerGas (The total must be the larger of the two)',
			)
		) {
			throw new InvalidGasLimitError(e.message, { cause: e })
		}
		throw new InternalError(e.message, { cause: e })
	}
	return /** @type {import('./ImpersonatedTx.js').ImpersonatedTx}*/ (
		new Proxy(tx, {
			get(target, prop) {
				if (prop === 'isImpersonated') {
					return true
				}
				if (prop === 'hash') {
					return () => {
						try {
							return target.hash()
						} catch (e) {
							return keccak256(target.getHashedMessageToSign(), 'bytes')
						}
					}
				}
				if (prop === 'isSigned') {
					return () => true
				}
				if (prop === 'getSenderAddress') {
					return () => txData.impersonatedAddress
				}
				return Reflect.get(target, prop)
			},
		})
	)
}
