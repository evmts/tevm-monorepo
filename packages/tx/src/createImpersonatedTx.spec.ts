import { Common } from '@ethereumjs/common'
import { FeeMarketEIP1559Transaction } from '@ethereumjs/tx'
import { InternalError, InvalidGasLimitError } from '@tevm/errors'
import { EthjsAddress } from '@tevm/utils'
import { type MockedFunction, afterEach, describe, expect, it, vi } from 'vitest'
import { createImpersonatedTx } from './createImpersonatedTx.js'

vi.mock('@ethereumjs/tx', async () => {
	const actualEthjsTx = (await vi.importActual('@ethereumjs/tx')) as any
	const mockClass = vi.fn()
	mockClass.mockImplementation((...args: any[]) => {
		return new actualEthjsTx.FeeMarketEIP1559Transaction(...args)
	})
	return {
		...actualEthjsTx,
		FeeMarketEIP1559Transaction: mockClass,
	}
})

const FeeMarketEIP1559TransactionMock = FeeMarketEIP1559Transaction as unknown as MockedFunction<any>

afterEach(async () => {
	vi.resetAllMocks()
	const actualEthjsTx = (await vi.importActual('@ethereumjs/tx')) as any
	FeeMarketEIP1559TransactionMock.mockImplementation((...args: any[]) => {
		return new actualEthjsTx.FeeMarketEIP1559Transaction(...args)
	})
})

describe(createImpersonatedTx.name, () => {
	it('should create an EIP-1559 tx impersonating the address', () => {
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		const tx = createImpersonatedTx({
			data,
			impersonatedAddress,
		})
		expect(tx.isImpersonated).toBe(true)
		expect(tx.hash()).toMatchSnapshot()
		expect(tx.getSenderAddress()).toEqual(impersonatedAddress)
		expect(tx.isSigned()).toBe(true)
	})

	it('should support Object.keys', () => {
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		expect(Object.keys(createImpersonatedTx({ impersonatedAddress }))).toMatchSnapshot()
	})

	it('should throw InternalError if EIP-1559 is not enabled', () => {
		const common = Common.custom(
			{},
			{
				eips: [],
			},
		)
		common.isActivatedEIP = (eip) => {
			if (eip === 1559) {
				return false
			}
			return true
		}
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		expect(() => createImpersonatedTx({ impersonatedAddress, data }, { common })).toThrow(
			new InternalError(
				'EIP-1559 is not enabled on Common. Tevm currently only supports 1559 and it should be enabled by default',
				{ cause: new Error('EIP-1559 not enabled on Common') },
			),
		)
	})

	it('should throw InvalidGasLimitError if bigger than MAX_INTEGER', () => {
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		const ethjsError = new Error(
			'gasLimit cannot exceed MAX_UINT64 (2^64-1), given 374144419156711147060143317175368453031918731001855 (tx type=2 hash=not available (unsigned) nonce=0 value=0 signed=false hf=error maxFeePerGas=undefined maxPriorityFeePerGas=undefined)',
		)
		FeeMarketEIP1559TransactionMock.mockImplementation(() => {
			throw ethjsError
		})
		expect(() => createImpersonatedTx({ impersonatedAddress, data, gasLimit: `0x${'ff'.repeat(21)}` })).toThrow(
			new InvalidGasLimitError(ethjsError.message, { cause: ethjsError }),
		)
	})

	it('should throw GasLimitExceededError if smaller than max priority fee', () => {
		const expectedError = new Error(
			'maxFeePerGas cannot be less than maxPriorityFeePerGas (The total must be the larger of the two)',
		)
		FeeMarketEIP1559TransactionMock.mockImplementation(() => {
			throw expectedError
		})
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		expect(() => createImpersonatedTx({ impersonatedAddress, data, maxFeePerGas: 0n })).toThrow(
			new InvalidGasLimitError(expectedError.message, { cause: expectedError }),
		)
	})

	it('should throw an error if FeeMarketEIP1559Transaction throws', () => {
		const expectedError = new Error('Constructor error')
		FeeMarketEIP1559TransactionMock.mockImplementation(() => {
			throw expectedError
		})

		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		expect(() => createImpersonatedTx({ impersonatedAddress, data })).toThrow(
			new InternalError(expectedError.message, { cause: expectedError }),
		)
	})

	it('should throw an error if FeeMarketEIP1559Transaction throws non error', () => {
		const notError = { not: 'error' }
		FeeMarketEIP1559TransactionMock.mockImplementation(() => {
			throw notError
		})
		const impersonatedAddress = EthjsAddress.fromString(`0x${'42'.repeat(20)}`)
		const data = '0x5234'
		expect(() => createImpersonatedTx({ impersonatedAddress, data })).toThrow(
			new InternalError('Unknown Error', { cause: notError }),
		)
	})
})
