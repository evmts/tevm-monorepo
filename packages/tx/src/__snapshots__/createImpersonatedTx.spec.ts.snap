// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createImpersonatedTx > should create an EIP-1559 tx impersonating the address 1`] = `
Uint8Array [
  157,
  170,
  99,
  19,
  41,
  61,
  59,
  28,
  44,
  237,
  138,
  45,
  178,
  77,
  231,
  121,
  60,
  7,
  132,
  194,
  115,
  97,
  230,
  210,
  22,
  119,
  226,
  186,
  190,
  131,
  101,
  221,
]
`;

exports[`createImpersonatedTx > should support Object.keys 1`] = `
[
  "cache",
  "activeCapabilities",
  "DEFAULT_CHAIN",
  "_type",
  "txOptions",
  "nonce",
  "gasLimit",
  "to",
  "value",
  "data",
  "v",
  "r",
  "s",
  "common",
  "chainId",
  "accessList",
  "AccessListJSON",
  "maxFeePerGas",
  "maxPriorityFeePerGas",
]
`;
